Allow to remove useless column from browser

Arthur-Milchior opened the pull request:
Currently, if for some reason, the browser have an active columns
which does not exists, instead of showing any information to the user
related to the column, it shows the name of the alphabetically first
column. This makes little sens. And worse, this column can't be
removed. Indeed, intuitively the user may want to click on the name
shown in the column to remove it. Instead it'll add the real first
column (or remove it if it really was here. In which case it only
removes the column with actual information.)

I thus do believe that it is slighly better to allow the user to
remove any useless columns (and not to add them back. At worst, the
add-on creating those columns will deal with them)
To do so, I add to show the "type" of the columns, since I don't have
access to their name.

Another solution which would makes sens too would simply be to hide
those columns from the users. Note that this would create potentially
a problem because in this case they may have 0 columns shown if all
active columns are columns anki does not know

To test this, you can easily open the debugger, and add faulty names to "activeCols".
E.g. 
`
from aqt import mw
mw.col.conf["activeCols"]+= ["faulty name", "other faulty name"]
`
open the browser and check that you can remove columns and not add them back (or you could do it the more natural way, using add-on adding columns and then disabling the add-on and restarting anki)

Arthur-Milchior commented:
Sorry if you received multiple notifications. I realized I pushed something on the wrong branch.

dae commented:
Is this a widespread issue, or is this problem limited to columns added by the advanced browser add-on? If it's the latter, perhaps the add-on is a better place to fix it.

Arthur-Milchior commented:
This is a problem related to columns added using the advanced browser, when you don't have the add-on anymore. If for some reason you delete the add-on or change the computer and don't want to install it (or even simply forget that you have installed it on an older computer), you get stuck with a bunch of "Answer" column which are actually empty and that you can't remove.
This is why I ensured that the code and the functionallity was changed as little as possible. To ensure that in this situaton: 
* anyone can understand that it's not actually "answer" column, but something which maybe makes slightly more sens if the user recall they used an add-on adding columns
* that they can remove those column if they want, and nothing else.

dae commented:
How about the following? If the add-on wants to preserve the data even when it's not active, it should be storing it separately and not requiring the core code to clean up after it.

```diff
--- a/aqt/browser.py
+++ b/aqt/browser.py
@@ -109,9 +109,6 @@ class DataModel(QAbstractTableModel):
                 if type == stype:
                     txt = name
                     break
-            # handle case where extension has set an invalid column type
-            if not txt:
-                txt = self.browser.columns[0][1]
             return txt
         else:
             return
@@ -550,6 +547,14 @@ class Browser(QMainWindow):
         ]
         self.columns.sort(key=itemgetter(1))

+    def removeUnsupportedColumns(self):
+        # clean up columns added by an add-on that's not active
+        valid = {x[0] for x in self.columns}
+        self.model.activeCols = [
+            x for x in self.model.activeCols
+            if x in valid
+        ]
+
     # Searching
     ######################################################################

@@ -636,6 +641,7 @@ class Browser(QMainWindow):
         self.form.tableView.selectionModel().selectionChanged.connect(self.onRowChanged)
         self.form.tableView.setStyleSheet("QTableView{ selection-background-color: rgba(127, 127, 127, 50);  }")
         self.singleCard = False
+        self.removeUnsupportedColumns()

     def setupEditor(self):
         self.editor = aqt.editor.Editor(
```

dae commented:
@hssm Perhaps when active, the add-on could store the columns in a separate "activeColsAdvanced"? The key could be factored into a constant to make it easy for the add-on to override.

Arthur-Milchior commented:
I think my answer was misleading. It's not the current version of "advanced browser" which creates the problem. I modified the add-on, and so that's why I didn't answer "no" to your question, even if I now understand that I should have done so. 

My point here was mostly that, even if the problem is probably rare, the way Anki deals with it seems quite confusing; I wanted to give a less confusing way to deal with it. In any case, it's not an important part of the code, and it solve what I believe to be a rare problem (in my cases, it occured only because I was editing an add-on, and added things in ['activeCols'] by accident, and I realized I couldn't easily remove those columns without using the debugger/editing the json in the database). 
Your way seems quite fine too.

dae commented:
Ok, if this is not affecting other users I don't think it makes sense to spend much time on it.

hssm commented:
For what it's worth, the add-on does keep its own columns under its own configuration key so it's unlikely a user will end up with "unofficial" columns in their column set anyway. The original `activeCols` is copied on collection open and restored on collection close. The add-on never pollutes the built-in data set with its own columns so removing the add-on has no real effect.
